contracts/Marketplace.sol

```diff
@@ -1,74 +1,296 @@
-pragma solidity ^0.4.24;
-
-import "openzeppelin-solidity/contracts/utils/Address.sol";
-import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
-import "openzeppelin-solidity/contracts/lifecycle/Pausable.sol";
-
-contract Service is Ownable, Pausable {
-    bytes32 public sid;
-    bytes32 public latest;
-    mapping (bytes32 => Version) public versions;
-    mapping (bytes32 => mapping(address => bool)) private _versionAccess;
-
-    struct Version {
-        bytes32 id;
-        bytes32 location;
-        uint256 price;
-        bool _exists;
-    }
+pragma solidity >=0.5.0 <0.6.0;
+
+import "../node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "../node_modules/openzeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "../node_modules/openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+
+contract Marketplace is Ownable, Pausable {
+  // ------------------------------------------------------
+  // Structures
+  // ------------------------------------------------------
+
+  struct Service {
+    address owner;
+    bytes sid;
+    uint256 price;
+
+    Version[] versions;
+    Payment[] payments;
+  }
+
+  struct Version {
+    bytes20 hash;
+    bytes url;
+  }
+
+  struct Payment {
+    address purchaser;
+  }
+
+  // ------------------------------------------------------
+  // State variables
+  // ------------------------------------------------------
+
+  Service[] public services;
+
+  IERC20 public tokenAddress;
+
+  // ------------------------------------------------------
+  // Constructor
+  // ------------------------------------------------------
+
+  constructor(IERC20 _tokenAddress) public {
+    tokenAddress = _tokenAddress;
+  }
+
+  // ------------------------------------------------------
+  // Events
+  // ------------------------------------------------------
+
+  event ServiceCreated(
+    uint indexed serviceIndex,
+    bytes sid,
```

Comment:
do we want sid to be indexed as well? (same question for other events)

Reply:
It will be better but it "transform" the types bytes (and bytes20) to a fixed length by adding 0s to the end of the hex. It pretty annoying to use.

Reply:
from https://solidity.readthedocs.io/en/v0.4.24/abi-spec.html#events

> For all fixed-length Solidity types, the EVENT_INDEXED_ARGS array contains the 32-byte encoded value directly. However, for types of dynamic length, which include string, bytes, and arrays, EVENT_INDEXED_ARGS will contain the Keccak hash of the encoded value, rather than the encoded value directly. This allows applications to efficiently query for values of dynamic-length types (by setting the hash of the encoded value as the topic), but leaves applications unable to decode indexed values they have not queried for. For dynamic-length types, application developers face a trade-off between fast search for predetermined values (if the argument is indexed) and legibility of arbitrary values (which requires that the arguments not be indexed). Developers may overcome this tradeoff and achieve both efficient search and arbitrary legibility by defining events with two arguments — one indexed, one not — intended to hold the same value.

So actually, indexed dynamic length type cannot get original data but the hash of it....

Reply:
What about duplicating indexed parameter with their non-indexed value?

Reply:
Hum I don't know need to read about this to find out the answer